pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,425 @@
+	cn.importConfigNodesChildren[stmt.Label] = NewImportConfigNode(stmt, childGlobals, sourceType)
+	return nil
+}
+
+// evaluateChildren evaluates the import nodes managed by this import node.
+func (cn *ImportConfigNode) evaluateChildren() error {
+	for _, child := range cn.importConfigNodesChildren {
+		err := child.Evaluate(&vm.Scope{
+			Parent:    nil,
+			Variables: make(map[string]interface{}),
+		})
+		if err != nil {
+			return fmt.Errorf("imported node %s failed to evaluate, %v", child.label, err)
+		}
+	}
+	return nil
+}
+
+// onChildrenContentUpdate notifies the parent that the content has been updated.
+func (cn *ImportConfigNode) onChildrenContentUpdate(child BlockNode) {
+	// If the node is already updating its content, it will call OnBlockNodeUpdate
+	// so the notification can be ignored.
+	if !cn.inContentUpdate.Load() {
+		cn.OnBlockNodeUpdate(cn)
+	}
+}
+
+// Run runs the managed source and the import children until ctx is
+// canceled. Evaluate must have been called at least once without returning an
+// error before calling Run.
+//
+// Run will immediately return ErrUnevaluated if Evaluate has never been called
+// successfully. Otherwise, Run will return nil.
+func (cn *ImportConfigNode) Run(ctx context.Context) error {
+	if cn.source == nil {
+		return ErrUnevaluated
+	}
+
+	newCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // This will stop the children and the managed source.
+
+	errChan := make(chan error, 1)
+
+	runner := runner.New(func(node *ImportConfigNode) runner.Worker {
+		return &childRunner{
+			node:    node,
+			errChan: errChan,
+		}
+	})
+	defer runner.Stop()
+
+	updateTasks := func() error {
+		cn.mut.Lock()
+		defer cn.mut.Unlock()
+		cn.importChildrenRunning = true
+		var tasks []*ImportConfigNode
+		for _, value := range cn.importConfigNodesChildren {
+			tasks = append(tasks, value)
+		}
+
+		return runner.ApplyTasks(newCtx, tasks)
+	}
+
+	err := updateTasks()
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import failed to run nested imports", "err", err)
+		cn.setRunHealth(component.HealthTypeExited, fmt.Sprintf("import shut down while trying to run nested imports with error: %s", err))
+		return err
+	}
+
+	go func() {
+		errChan <- cn.source.Run(newCtx)
+	}()
+
+	cn.setRunHealth(component.HealthTypeHealthy, "started import")
+	err = cn.run(errChan, updateTasks)
+
+	var exitMsg string
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import exited with error", "err", err)
+		exitMsg = fmt.Sprintf("import shut down with error: %s", err)
+	} else {
+		level.Info(cn.logger).Log("msg", "import exited")
+		exitMsg = "import shut down normally"
+	}
+	cn.setRunHealth(component.HealthTypeExited, exitMsg)
+	return err
+}
+
+func (cn *ImportConfigNode) run(errChan chan error, updateTasks func() error) error {
+	for {
+		select {
+		case <-cn.importChildrenUpdateChan:
+			err := updateTasks()
+			if err != nil {
+				return err
+			}
+		case err := <-errChan:
+			return err
```

Comment:
The Flow controller expects that the `Run` method only exits in two scenarios:

1. The provided context is cancelled (in which case no error should be returned) 
2. A fatal unrecoverable error occurs, where calling `Run` again would not resolve it. 

Because of these assumptions, the controller will never call Run again for the same node once it returns. Returning early here is signaling a fatal error, but I'm not sure that it really is one for all writes to run. 

Can you put more thought into the failure modes and try to reduce the blast radius of failure here?